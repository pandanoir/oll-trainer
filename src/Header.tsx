import { faBars, faTimes } from '@fortawesome/free-solid-svg-icons';
import { useState, VFC } from 'react';
import { createPortal } from 'react-dom';
import { NavLink } from 'react-router-dom';
import tw from 'twin.macro';

import { IconButton } from './components/common/IconButton';
import { RouteList } from './route';
import { usePortalRoot } from './utils/hooks/usePortalRoot';

const SideMenu: VFC<{ hidden?: boolean; onClose: () => void }> = ({
  hidden = false,
  onClose,
}) =>
  createPortal(
    <div
      css={[
        hidden
          ? tw`hidden`
          : tw`md:hidden absolute z-10 inset-0 bg-gray-300 bg-opacity-30 dark:bg-black dark:bg-opacity-50`,
      ]}
      onClick={onClose}
    >
      <div
        onClick={(event) => event.stopPropagation()}
        tw="m-0 p-3 pr-6 flex flex-col items-start space-y-3 md:hidden list-none border-r-2 absolute top-0 bottom-0 bg-white dark:bg-gray-800"
      >
        <IconButton tw="dark:text-white" icon={faTimes} onClick={onClose} />
        <ul tw="flex flex-col space-y-1">
          {RouteList.map(({ path, name }) => (
            <li key={path}>
              <NavLink
                to={path}
                exact
                tw="no-underline text-gray-600 dark:text-gray-300"
                activeClassName="font-bold"
                onClick={onClose}
              >
                {name}
              </NavLink>
            </li>
          ))}
        </ul>
      </div>
    </div>,
    usePortalRoot().current
  );

export const Header: VFC<{ right: JSX.Element }> = ({ right }) => {
  const [isMenuVisible, setIsMenuVisible] = useState(false);
  return (
    <header tw="px-3 pt-5 pb-3 bg-white dark:bg-gray-800">
      <nav tw="flex place-content-between">
        <span>
          <IconButton
            icon={faBars}
            tw="md:hidden"
            onClick={() => setIsMenuVisible((b) => !b)}
          />
          <ul tw="m-0 p-0 list-none hidden md:flex md:flex-row md:space-x-3">
            {RouteList.map(({ path, name }) => (
              <li key={path}>
                <NavLink
                  to={path}
                  exact
                  tw="no-underline text-gray-600 dark:text-gray-300"
                  activeClassName="font-bold"
                >
                  {name}
                </NavLink>
              </li>
            ))}
          </ul>
        </span>
        {right}
      </nav>
      <SideMenu
        hidden={!isMenuVisible}
        onClose={() => setIsMenuVisible(false)}
      />
    </header>
  );
};
