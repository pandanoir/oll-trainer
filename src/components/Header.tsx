import {
  faBars,
  faExternalLinkAlt,
  faTimes,
} from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { useState, VFC } from 'react';
import { createPortal } from 'react-dom';
import { NavLink } from 'react-router-dom';
import tw from 'twin.macro';

import { RouteList } from '../route';
import { usePortalRoot } from '../utils/hooks/usePortalRoot';
import { IconButton } from './common/IconButton';

const SideMenu: VFC<{ hidden?: boolean; onClose: () => void }> = ({
  hidden = false,
  onClose,
}) => {
  const [prevHidden, setPrevHidden] = useState(hidden);

  return createPortal(
    <div
      css={[
        hidden && prevHidden
          ? tw`opacity-0 pointer-events-none`
          : tw`md:hidden absolute z-10 inset-0 bg-gray-300 bg-opacity-30 dark:bg-black dark:bg-opacity-50`,
      ]}
      onClick={onClose}
    >
      <div
        onClick={(event) => event.stopPropagation()}
        css={[
          tw`m-0 p-3 pr-6 flex flex-col transform items-start space-y-3 md:hidden list-none border-r-2 absolute top-0 bottom-0 bg-white dark:bg-gray-800 transition-transform duration-150`,
          hidden ? tw`-translate-x-full` : tw`translate-x-0`,
        ]}
        onTransitionEnd={() => setPrevHidden(hidden)}
      >
        <IconButton
          tw="dark:text-white p-1 py-0.5 text-lg"
          icon={faTimes}
          onClick={onClose}
        />
        <ul tw="flex flex-col space-y-1">
          {RouteList.map(({ path, name }) => (
            <li key={path}>
              <NavLink
                to={path}
                exact
                tw="inline-block no-underline text-gray-600 dark:text-gray-300 py-1.5"
                activeClassName="font-bold"
                onClick={onClose}
              >
                {name}
              </NavLink>
            </li>
          ))}
          <li>
            <a
              href="https://github.com/pandanoir/oll-trainer"
              target="_blank"
              rel="noreferrer"
              tw="no-underline text-gray-600 dark:text-gray-300 flex space-x-1 items-center"
            >
              <span>GitHub</span>
              <FontAwesomeIcon tw="text-sm" icon={faExternalLinkAlt} />
            </a>
          </li>
        </ul>
      </div>
    </div>,
    usePortalRoot().current
  );
};

export const Header: VFC<{ right: JSX.Element }> = ({ right }) => {
  const [isMenuVisible, setIsMenuVisible] = useState(false);
  return (
    <header tw="px-3 pt-5 pb-3 bg-white dark:bg-gray-800">
      <nav tw="flex place-content-between">
        <span>
          <IconButton
            icon={faBars}
            tw="md:hidden"
            onClick={() => setIsMenuVisible((b) => !b)}
          />
          <ul tw="m-0 p-0 list-none hidden md:flex md:flex-row md:space-x-3">
            {RouteList.map(({ path, name }) => (
              <li key={path}>
                <NavLink
                  to={path}
                  exact
                  tw="no-underline text-gray-600 dark:text-gray-300"
                  activeClassName="font-bold"
                >
                  {name}
                </NavLink>
              </li>
            ))}
            <li>
              <a
                href="https://github.com/pandanoir/oll-trainer"
                target="_blank"
                rel="noreferrer"
                tw="no-underline text-gray-600 dark:text-gray-300 flex space-x-1 items-center"
              >
                <span>GitHub</span>
                <FontAwesomeIcon tw="text-sm" icon={faExternalLinkAlt} />
              </a>
            </li>
          </ul>
        </span>
        {right}
      </nav>
      <SideMenu
        hidden={!isMenuVisible}
        onClose={() => setIsMenuVisible(false)}
      />
    </header>
  );
};
