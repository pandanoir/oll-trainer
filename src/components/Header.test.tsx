/**
 * @jest-environment jsdom
 */
import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { Header } from './Header';

jest.mock('../pages/CpPage', () => ({ CpPage: () => null }));
jest.mock('../pages/EoQuizPage', () => ({ EoQuizPage: () => null }));
jest.mock('../pages/InspectionPage', () => ({ InspectionPage: () => null }));
jest.mock('../pages/LearningPage', () => ({ LearningPage: () => null }));
jest.mock('../pages/OllPage', () => ({ OllPage: () => null }));
jest.mock('../pages/ScramblePage', () => ({ ScramblePage: () => null }));
jest.mock('../pages/TimerPage', () => ({ TimerPage: () => null }));

describe('Header', () => {
  it('snapshot', () => {
    const { asFragment } = render(
      <BrowserRouter>
        <Header onSettingButtonClick={jest.fn()} onDarkModeToggle={jest.fn()} />
      </BrowserRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
