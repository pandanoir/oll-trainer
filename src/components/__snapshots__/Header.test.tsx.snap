// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header snapshot 1`] = `
<DocumentFragment>
  <header
    class="_StyledHeader-sc-fsl829-8 ceERtR"
  >
    <nav
      class="_StyledNav-sc-fsl829-9 dQatxH"
    >
      <span>
        <button
          class="_StyledIconButton2-sc-fsl829-10 itbLKO"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-bars "
            data-icon="bars"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 96C0 78.33 14.33 64 32 64H416C433.7 64 448 78.33 448 96C448 113.7 433.7 128 416 128H32C14.33 128 0 113.7 0 96zM0 256C0 238.3 14.33 224 32 224H416C433.7 224 448 238.3 448 256C448 273.7 433.7 288 416 288H32C14.33 288 0 273.7 0 256zM416 448H32C14.33 448 0 433.7 0 416C0 398.3 14.33 384 32 384H416C433.7 384 448 398.3 448 416C448 433.7 433.7 448 416 448z"
              fill="currentColor"
            />
          </svg>
        </button>
        <ul
          class="_StyledUl2-sc-fsl829-11 cTvaRg"
        >
          <li>
            <a
              class="_StyledCustomNavLink2-sc-fsl829-0 bSPkuj"
              href="/timer"
            >
              Hi-Timer
            </a>
          </li>
          <li>
            <a
              class="_StyledCustomNavLink2-sc-fsl829-0 bSPkuj"
              href="/oll"
            >
              oll
            </a>
          </li>
          <li>
            <a
              class="_StyledCustomNavLink2-sc-fsl829-0 bSPkuj"
              href="/learn"
            >
              learn
            </a>
          </li>
          <li>
            <a
              class="_StyledCustomNavLink2-sc-fsl829-0 bSPkuj"
              href="/scramble"
            >
              scramble
            </a>
          </li>
          <li>
            <a
              class="_StyledCustomNavLink2-sc-fsl829-0 bSPkuj"
              href="/inspection"
            >
              inspection
            </a>
          </li>
          <li>
            <a
              class="_StyledCustomNavLink2-sc-fsl829-0 bSPkuj"
              href="/cp"
            >
              cp check
            </a>
          </li>
          <li>
            <a
              class="_StyledCustomNavLink2-sc-fsl829-0 bSPkuj"
              href="/quiz"
            >
              quiz
            </a>
          </li>
          <li>
            <a
              class="_StyledA2-sc-fsl829-12 hpqlLB"
              href="https://github.com/pandanoir/oll-trainer"
              rel="noreferrer"
              target="_blank"
            >
              <span>
                GitHub
              </span>
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-up-right-from-square _StyledFontAwesomeIcon2-sc-fsl829-13 eHZezK"
                data-icon="up-right-from-square"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M384 320c-17.67 0-32 14.33-32 32v96H64V160h96c17.67 0 32-14.32 32-32s-14.33-32-32-32L64 96c-35.35 0-64 28.65-64 64V448c0 35.34 28.65 64 64 64h288c35.35 0 64-28.66 64-64v-96C416 334.3 401.7 320 384 320zM488 0H352c-12.94 0-24.62 7.797-29.56 19.75c-4.969 11.97-2.219 25.72 6.938 34.88L370.8 96L169.4 297.4c-12.5 12.5-12.5 32.75 0 45.25C175.6 348.9 183.8 352 192 352s16.38-3.125 22.62-9.375L416 141.3l41.38 41.38c9.156 9.141 22.88 11.84 34.88 6.938C504.2 184.6 512 172.9 512 160V24C512 10.74 501.3 0 488 0z"
                  fill="currentColor"
                />
              </svg>
            </a>
          </li>
        </ul>
      </span>
      <span
        class="_StyledSpan-sc-fsl829-14 fEwigp"
      >
        <button
          class="_StyledIconButton3-sc-fsl829-15 frJWXH"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-gear "
            data-icon="gear"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M495.9 166.6C499.2 175.2 496.4 184.9 489.6 191.2L446.3 230.6C447.4 238.9 448 247.4 448 256C448 264.6 447.4 273.1 446.3 281.4L489.6 320.8C496.4 327.1 499.2 336.8 495.9 345.4C491.5 357.3 486.2 368.8 480.2 379.7L475.5 387.8C468.9 398.8 461.5 409.2 453.4 419.1C447.4 426.2 437.7 428.7 428.9 425.9L373.2 408.1C359.8 418.4 344.1 427 329.2 433.6L316.7 490.7C314.7 499.7 307.7 506.1 298.5 508.5C284.7 510.8 270.5 512 255.1 512C241.5 512 227.3 510.8 213.5 508.5C204.3 506.1 197.3 499.7 195.3 490.7L182.8 433.6C167 427 152.2 418.4 138.8 408.1L83.14 425.9C74.3 428.7 64.55 426.2 58.63 419.1C50.52 409.2 43.12 398.8 36.52 387.8L31.84 379.7C25.77 368.8 20.49 357.3 16.06 345.4C12.82 336.8 15.55 327.1 22.41 320.8L65.67 281.4C64.57 273.1 64 264.6 64 256C64 247.4 64.57 238.9 65.67 230.6L22.41 191.2C15.55 184.9 12.82 175.3 16.06 166.6C20.49 154.7 25.78 143.2 31.84 132.3L36.51 124.2C43.12 113.2 50.52 102.8 58.63 92.95C64.55 85.8 74.3 83.32 83.14 86.14L138.8 103.9C152.2 93.56 167 84.96 182.8 78.43L195.3 21.33C197.3 12.25 204.3 5.04 213.5 3.51C227.3 1.201 241.5 0 256 0C270.5 0 284.7 1.201 298.5 3.51C307.7 5.04 314.7 12.25 316.7 21.33L329.2 78.43C344.1 84.96 359.8 93.56 373.2 103.9L428.9 86.14C437.7 83.32 447.4 85.8 453.4 92.95C461.5 102.8 468.9 113.2 475.5 124.2L480.2 132.3C486.2 143.2 491.5 154.7 495.9 166.6V166.6zM256 336C300.2 336 336 300.2 336 255.1C336 211.8 300.2 175.1 256 175.1C211.8 175.1 176 211.8 176 255.1C176 300.2 211.8 336 256 336z"
              fill="currentColor"
            />
          </svg>
        </button>
        <span
          class="_StyledSpan2-sc-fsl829-16 cZUbTj"
        >
          <button
            class="_StyledButton-sc-15dm4so-0 cvkjJj"
          >
            <div
              class="_StyledDiv-sc-15dm4so-1 itfcfF"
            />
          </button>
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-moon "
            data-icon="moon"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M32 256c0-123.8 100.3-224 223.8-224c11.36 0 29.7 1.668 40.9 3.746c9.616 1.777 11.75 14.63 3.279 19.44C245 86.5 211.2 144.6 211.2 207.8c0 109.7 99.71 193 208.3 172.3c9.561-1.805 16.28 9.324 10.11 16.95C387.9 448.6 324.8 480 255.8 480C132.1 480 32 379.6 32 256z"
              fill="currentColor"
            />
          </svg>
        </span>
      </span>
    </nav>
  </header>
</DocumentFragment>
`;
